File: fastapi_backend/services/embeddings.py
from functools import lru_cache
from typing import List
from google import genai
from config import get_settings


@lru_cache()
def get_gemini_client() -> genai.Client:
    settings = get_settings()
    if not settings.gemini_api_key:
        raise ValueError("GEMINI_API_KEY is required for AI features")
    return genai.Client(api_key=settings.gemini_api_key)


def to_pg_vector(values: List[float]) -> str:
    """Convert float list to PostgreSQL vector literal string."""
    return "[" + ",".join(str(v) for v in values) + "]"


async def embed_document(text: str, title: str = "none") -> List[float]:
    return await _embed_text(f"title: {title or 'none'} | text: {text}")


async def embed_query(text: str) -> List[float]:
    return await _embed_text(f"task: question answering | query: {text}")


async def _embed_text(text: str) -> List[float]:
    import asyncio
    settings = get_settings()
    client = get_gemini_client()

    loop = asyncio.get_event_loop()
    response = await loop.run_in_executor(
        None,
        lambda: client.models.embed_content(
            model=settings.gemini_embedding_model,
            contents=text,
            config={"output_dimensionality": 1536},
        ),
    )

    values = response.embeddings[0].values if response.embeddings else None
    if not values:
        raise ValueError("Gemini did not return an embedding.")
    return list(values)

File: FRONTEND_NEXTJS/src/lib/embeddings.ts
import { GoogleGenAI } from "@google/genai";

let cachedGemini: GoogleGenAI | null = null;

export function getGeminiClient() {
  if (!process.env.GEMINI_API_KEY) {
    throw new Error("GEMINI_API_KEY is required for AI features");
  }

  if (!cachedGemini) {
    cachedGemini = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  }

  return cachedGemini;
}

export async function embedDocument(text: string, title = "none") {
  return embedText(`title: ${title || "none"} | text: ${text}`);
}

export async function embedQuery(text: string) {
  return embedText(`task: question answering | query: ${text}`);
}

async function embedText(text: string) {
  const client = getGeminiClient();
  const response = await client.models.embedContent({
    model: process.env.GEMINI_EMBEDDING_MODEL || "gemini-embedding-2",
    contents: text,
    config: { outputDimensionality: 1536 }
  });

  const values = response.embeddings?.[0]?.values;
  if (!values?.length) {
    throw new Error("Gemini did not return an embedding.");
  }

  return values;
}

export function toPgVector(values: number[]) {
  return `[${values.join(",")}]`;
}
